packages/@sanity/preview/src/observeFields.js

```diff
@@ -61,18 +48,41 @@ type Cache = {[id: Id]: CachedFieldObserver[]}
 const CACHE: Cache = {} // todo: use a LRU cache instead (e.g. hashlru or quick-lru)
 
 function createCachedFieldObserver(id, fields): CachedFieldObserver {
-  let latest = null
+  let cachedLatest = null
   const changes$ = new Observable(observer => {
-    if (latest) {
-      // Re-emit last known value immediately
-      observer.next(latest)
-      return fetchDocumentPathsSlow(id, fields)
-        .concat(listen(id).switchMap(event => fetchDocumentPathsSlow(id, fields)))
-        .subscribe(observer)
-    }
-    return listenFields(id, fields).subscribe(observer)
+    observer.next(cachedLatest)
+    observer.complete()
   })
-    .do(v => (latest = v))
+    .mergeMap(latest => {
+      const snapshotEvents = latest // Re-emit last known value immediately
+        ? Observable.of(latest)
+            .concat(fetchDocumentPathsSlow(id, fields))
+            .map(snapshot => ({type: 'snapshot', snapshot: snapshot}))
+        : fetchDocumentPathsFast(id, fields).mergeMap(
+            result =>
+              result === undefined
+                ? // hack: if we get undefined as result here it is most likely because the document has
+                  // just been created and is not yet indexed. We therefore need to wait a bit and then re-fetch.
+                  fetchDocumentPathsSlow(id, fields)
+                : Observable.of(result)
+          )
+      return snapshotEvents
+        .map(snapshot => ({type: 'snapshot', snapshot: snapshot}))
+        .concat(listen(id))
+    })
+    .scan((prevSnapshot, event) => {
+      if (event.type === 'snapshot') {
+        return event.snapshot
+      }
+      if (event.type === 'mutation') {
+        return applyMutations(prevSnapshot, event.mutations)
+      }
+      // eslint-disable-next-line no-console
+      console.warn(new Error(`Invalid event: ${event.type}`))
+      return null
+    }, null)
+    .filter(Boolean)
```

Comment:
Vil denne gjøre at meldinger om dokumenter som er blitt slettet (applyMutation returnerer null) forsvinner?

Reply:
Yeah, good catch! I should rather filter out unknown events before entering the scan operator above.